### PR TITLE
Migrate the CT map during upgrade

### PR DESCRIPTION
**Pull request by joestringer**
Issue #5070 revealed a case during upgrade/downgrade where the CT map might not be recreated with the correct properties, which could lead to userspace opening and dumping the kernel version of the map and overwriting random memory in the userspace process.

To avoid this, during endpoint restore check if the CT map pinned to the filesystem has properties that mismatch the current properties for the CT map (as defined in the current code's definitions). If so, remove the map.

When the map is removed from the filesystem, existing BPF programs that make use of the map will continue to operate correctly on the version of the map that has been unpinned from the filesystem. The next time an endpoint is regenerated, it will create a new version of the CT map. All endpoints are expected to be regenerated during Cilium restart, and therefore will migrate naturally to the new CT map.

This is performed for both global and local conntrack maps.

#### Upgrade impact

Note that this may cause temporary disruption of ongoing TCP connections during upgrade from Cilium 1.1 or earlier to Cilium 1.2 or later. Connections should recover shortly after upgrade.

This solution must be present in the target version of the code, that is to say, if a user intends to safely downgrade from Cilium 1.2 to an earlier version:
* To Cilium 1.0: Use version 1.0.6 or later
* To Cilium 1.1: Use version 1.1.3 or later

(Note that at the time of this PR the above releases are the N+1 target micro releases, and are not yet available)

Fixes: #5070

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5118)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
One thing that needs to be checked: is it guaranteed that the CT map is not dumped prior to this code running?

EDIT: Yes.

Endpoint restore occurs in `NewDaemon()`, which is the first thing run by `runDaemon()`. `runDaemon()` subsequently starts the CT garbage collector only if `NewDaemon()` successfully initialized.

**Comment by joestringer** [comment]
Failure is a panic:

```
Aug 07 15:04:54 runtime cilium-agent[7322]: level=info msg="Launching Cilium health endpoint" subsys=daemon
Aug 07 15:04:54 runtime cilium-agent[7322]: level=debug msg="Starting new controller" name=cilium-health-ep subsys=controller uuid=3d8c3131-9a53-11e8-a9c2-0800275579e8
Aug 07 15:04:54 runtime cilium-agent[7322]: level=info msg="Enabling docker event listener" subsys=workload-watcher
Aug 07 15:04:54 runtime cilium-agent[7322]: panic: runtime error: invalid memory address or nil pointer dereference
Aug 07 15:04:54 runtime cilium-agent[7322]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1e37f42]
Aug 07 15:04:54 runtime cilium-agent[7322]: goroutine 996 [running]:
Aug 07 15:04:54 runtime cilium-agent[7322]: github.com/cilium/cilium/pkg/endpoint.(*Endpoint).StringID(0x0, 0x0, 0xc421cf16e0)
Aug 07 15:04:54 runtime cilium-agent[7322]: 	/home/vagrant/go/src/github.com/cilium/cilium/pkg/endpoint/endpoint.go:1236 +0x22
Aug 07 15:04:54 runtime cilium-agent[7322]: github.com/cilium/cilium/pkg/maps/ctmap.GetMapPath(0x293e760, 0x0, 0xc421365c01, 0x1, 0x0)
Aug 07 15:04:54 runtime cilium-agent[7322]: 	/home/vagrant/go/src/github.com/cilium/cilium/pkg/maps/ctmap/ctmap.go:91 +0x57
Aug 07 15:04:54 runtime cilium-agent[7322]: github.com/cilium/cilium/pkg/endpointmanager.RunGC(0x0, 0x1, 0xc421365f98)
Aug 07 15:04:54 runtime cilium-agent[7322]: 	/home/vagrant/go/src/github.com/cilium/cilium/pkg/endpointmanager/conntrack.go:46 +0x79
Aug 07 15:04:54 runtime cilium-agent[7322]: github.com/cilium/cilium/pkg/endpointmanager.EnableConntrackGC.func1(0xc420290cf0, 0x101)
Aug 07 15:04:54 runtime cilium-agent[7322]: 	/home/vagrant/go/src/github.com/cilium/cilium/pkg/endpointmanager/conntrack.go:80 +0x29d
Aug 07 15:04:54 runtime cilium-agent[7322]: created by github.com/cilium/cilium/pkg/endpointmanager.EnableConntrackGC
Aug 07 15:04:54 runtime cilium-agent[7322]: 	/home/vagrant/go/src/github.com/cilium/cilium/pkg/endpointmanager/conntrack.go:70 +0x69
```

**Comment by joestringer** [comment]
Unexpected aspect of golang, if you pass a nil pointer into a function which subsequently passes that pointer into a function that takes an interface as an argument, then the pointer is no longer nil when passed as an interface:
https://play.golang.org/p/xKc_QK73sXk

Checking and directly passing nil from `RunGC()` fixes the issue.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Key size calculation was wrong (kernel rounds the size up to nearest multiple of 8), and there was a double-delete of the map file which was being treated as fatal. Fixed these issues now, should be good.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Hit #5121:

https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/5498/testReport/junit/k8s-1/11/K8sKafkaPolicyTest_Kafka_Policy_Tests_KafkaPolicies/

https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/5498/testReport/junit/k8s-1/11/K8sKafkaPolicyTest_Kafka_Policy_Tests_KafkaPolicies/attachments/65adb2a0_K8sKafkaPolicyTest_Kafka_Policy_Tests_KafkaPolicies.zip

```
 [2018-08-08 13:54:40,759] INFO Initiating client connection, connectString=zook:2181 sessionTimeout=60000 watcher=kafka.zookeeper.ZooKeeperClient$ZooKeeperClientWatcher$@be64738 (org.apache.zookeeper.ZooKeeper)
 [2018-08-08 13:54:40,787] ERROR Fatal error during KafkaServer startup. Prepare to shutdown (kafka.server.KafkaServer)
 java.net.UnknownHostException: zook: Name does not resolve
        at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
        at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928)
        at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323)
        at java.net.InetAddress.getAllByName0(InetAddress.java:1276)
        at java.net.InetAddress.getAllByName(InetAddress.java:1192)
        at java.net.InetAddress.getAllByName(InetAddress.java:1126)
        at org.apache.zookeeper.client.StaticHostProvider.<init>(StaticHostProvider.java:61)
        at org.apache.zookeeper.ZooKeeper.<init>(ZooKeeper.java:445)
        at org.apache.zookeeper.ZooKeeper.<init>(ZooKeeper.java:380)
        at kafka.zookeeper.ZooKeeperClient.<init>(ZooKeeperClient.scala:86)
        at kafka.zk.KafkaZkClient$.apply(KafkaZkClient.scala:1539)
        at kafka.server.KafkaServer.createZkClient$1(KafkaServer.scala:345)
        at kafka.server.KafkaServer.initZkClient(KafkaServer.scala:369)
        at kafka.server.KafkaServer.startup(KafkaServer.scala:202)
        at kafka.server.KafkaServerStartable.startup(KafkaServerStartable.scala:38)
        at kafka.Kafka$.main(Kafka.scala:92)
        at kafka.Kafka.main(Kafka.scala)
```

**Comment by joestringer** [comment]
test-me-please

File: pkg/bpf/map.go [hunk 1]
```diff
@@ -867,3 +867,29 @@ func (m *Map) resolveErrors() error {
...
 	return nil
 }
+
+// KeySize determines the BPF map key size for the specified size, by rounding
+// up the size according to the rules used by the kernel.
+//
+// Usage: bpf.KeySize(unsafe.Sizeof(foo{}))
+func KeySize(size uintptr) uint32 {
+	rem := size % 8
+	return uint32(size + (8 - rem))
```
Comment (joestringer) [inline pkg/bpf/map.go]: Without this logic, I was seeing the IPv6 CT map being removed/upgraded on startup (size new=38, old=40), but with this logic it seems like the IPv4 CT map is being removed/upgraded (size new=16, old=14). Need to dig into whether this is strictly necessary or not.

Comment (joestringer) [reply pkg/bpf/map.go]: Turns out the IPv4 CT tuple key is backed on the BPF side, but the IPv6 one isn't. @borkmann any preference on packed vs. not?

**Comment by joestringer** [comment]
`k8s-1.7.K8sServicesTest Bookinfo Demo Tests bookinfo demo` failed, providing not terribly useful logs:

[1663dc54_K8sServicesTest_Bookinfo_Demo_Tests_bookinfo_demo.zip](https://github.com/cilium/cilium/files/2271305/1663dc54_K8sServicesTest_Bookinfo_Demo_Tests_bookinfo_demo.zip)

Seems like it failed to regenerate an endpoint within a minute because load is too high. Pprof is also running at the time, hit #5135.

---

`k8s-1.7.K8sChaosTest Endpoint can still connect while Cilium is not running` also failed:

[0a9ca6ca_K8sChaosTest_Endpoint_can_still_connect_while_Cilium_is_not_running.zip](https://github.com/cilium/cilium/files/2271312/0a9ca6ca_K8sChaosTest_Endpoint_can_still_connect_while_Cilium_is_not_running.zip)

Preflight check failed, seems like a connectivity issue from Cilium to k8s. Following controller had a failure:

```
  sync-cnp-policy-status (v2 default/cnp-second-namespace)      35m36s ago     never        0       no error                    
  sync-cnp-policy-status (v2 default/cnp-specs)                 never          1m40s ago    16      context deadline exceeded   
```

Lots of complaints about not being able to communicate with what I assume to be the k8s apiserver, eg:

```
pod-kube-system-cilium-23lbr-cilium-agent.log:2018-08-08T16:39:54.00741574Z level=error msg="Cannot update CEP" containerID=d7ba028866 controller="sync-to-k8s-ciliumendpoint (23947)" endpointID=23947 error
="etcdserver: request timed out" ipv4=10.10.1.110 ipv6="f00d::a0a:100:0:5d8b" k8sPodName=/ policyRevision=0
...
pod-kube-system-cilium-23lbr-cilium-agent.log:2018-08-08T16:43:11.70060364Z level=error msg=k8sError error="github.com/cilium/cilium/daemon/k8s_watcher.go:497: Failed to list *v1.Pod: Get https://10.96.0.1
:443/api/v1/pods?limit=500&resourceVersion=0: dial tcp 10.96.0.1:443: connect: connection refused" subsys=daemon
...
pod-kube-system-cilium-23lbr-cilium-agent.log:2018-08-08T16:43:15.706049795Z level=error msg="Cannot get CEP for update" containerID=d7ba028866 controller="sync-to-k8s-ciliumendpoint (23947)" endpointID=23
947 error="Get https://10.96.0.1:443/apis/cilium.io/v2/namespaces/default/ciliumendpoints/reviews-v1-1217721402-rcdn6: dial tcp 10.96.0.1:443: connect: connection refused" ipv4=10.10.1.110 ipv6="f00d::a0a:
100:0:5d8b" k8sPodName=/ policyRevision=0

```

On further investigation of the kube-apiserver logs, I see #5085 .

**Comment by joestringer** [comment]
test-me-please

File: pkg/bpf/map.go (hunk 1 above)
Comment (joestringer) [reply pkg/bpf/map.go]: I opted to pack the ipv6 tuple, since that seems to make the C definitions of the key match the Golang definitions.

**Comment by joestringer** [comment]
Hit #5065:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/5523/testReport/junit/k8s-1/7/K8sPolicyTest_Basic_Test_checks_all_kind_of_Kubernetes_policies/

```
$ grep -C 1  ab74c810f1764e69ef89 kubelet-k8s1-1.7.log
Aug 09 10:01:19 k8s1 kubelet[3519]: I0809 10:01:19.804011    3519 kuberuntime_manager.go:466] Container {Name:empire-backup Image:docker.io/cilium/kafkaclient Command:[] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[] Resources:{Limits:map[] Requests:map[]} VolumeMounts:[{Name:default-token-mltnm ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath:}] LivenessProbe:nil ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:Always SecurityContext:nil Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
Aug 09 10:01:19 k8s1 kubelet[3519]: W0809 10:01:19.947825    3519 pod_container_deletor.go:77] Container "ab74c810f1764e69ef899e9083b642eda63a831ed5e36a795c7a0d24f918a43b" not found in pod's containers
Aug 09 10:01:20 k8s1 kubelet[3519]: level=debug msg="Processing CNI ADD request" args="&{ab74c810f1764e69ef899e9083b642eda63a831ed5e36a795c7a0d24f918a43b /proc/15428/ns/net eth0 IgnoreUnknown=1;K8S_POD_NAMESPACE=default;K8S_POD_NAME=empire-hq-2917507631-s1xbr;K8S_POD_INFRA_CONTAINER_ID=ab74c810f1764e69ef899e9083b642eda63a831ed5e36a795c7a0d24f918a43b /opt/cilium-cni/bin:/opt/cni/bin [123 34 99 110 105 86 101 114 115 105 111 110 34 58 34 34 44 34 110 97 109 101 34 58 34 99 105 108 105 117 109 34 44 34 116 121 112 101 34 58 34 99 105 108 105 117 109 45 99 110 105 34 125]}" eventUUID=2a7f2767-9bbb-11e8-904e-0800275579e8 subsys=cilium-cni
Aug 09 10:01:21 k8s1 kubelet[3519]: level=debug msg="Created veth pair" subsys=endpoint-connector vethPair="[tmpab74c lxcab74c]"
--
Aug 09 10:02:14 k8s1 kubelet[3519]: W0809 10:02:14.704876    3519 helpers.go:793] eviction manager: no observation found for eviction signal allocatableNodeFs.available
Aug 09 10:02:16 k8s1 kubelet[3519]: level=warning msg="Unable to create endpoint" containerID=ab74c810f1764e69ef899e9083b642eda63a831ed5e36a795c7a0d24f918a43b endpointID=23454 error="[PUT /endpoint/{id}][500] putEndpointIdFailed  endpoint 23454 did not synchronously regenerate after timeout" eventUUID=2a7f2767-9bbb-11e8-904e-0800275579e8 subsys=cilium-cni
Aug 09 10:02:16 k8s1 kubelet[3519]: E0809 10:02:16.492050    3519 cni.go:294] Error adding network: Unable to create endpoint: [PUT /endpoint/{id}][500] putEndpointIdFailed  endpoint 23454 did not synchronously regenerate after timeout
--
Aug 09 10:02:16 k8s1 kubelet[3519]: E0809 10:02:16.923824    3519 pod_workers.go:182] Error syncing pod 276ea6b1-9bbb-11e8-8aea-0800275579e8 ("empire-hq-2917507631-s1xbr_default(276ea6b1-9bbb-11e8-8aea-0800275579e8)"), skipping: failed to "CreatePodSandbox" for "empire-hq-2917507631-s1xbr_default(276ea6b1-9bbb-11e8-8aea-0800275579e8)" with CreatePodSandboxError: "CreatePodSandbox for pod \"empire-hq-2917507631-s1xbr_default(276ea6b1-9bbb-11e8-8aea-0800275579e8)\" failed: rpc error: code = 2 desc = NetworkPlugin cni failed to set up pod \"empire-hq-2917507631-s1xbr_default\" network: Unable to create endpoint: [PUT /endpoint/{id}][500] putEndpointIdFailed  endpoint 23454 did not synchronously regenerate after timeout"
```

**Comment by joestringer** [comment]
Hit also #5152.

**Comment by joestringer** [comment]
test-me-please

**Review by aanm — CHANGES_REQUESTED** [review]
Shouldn't we advise users to upgrade to 1.0.6 or 1.1.3 before upgrading to the next minor version in the docs?

**Comment by joestringer** [comment]
@aanm I agree, though we don't have any overview section at the moment where this would fit well. I had a chat with some people in Slack a few weeks ago about how we could improve the workflow of the upgrade docs, and I think it would be good to separately take a look at this for 1.2. I'll see if I can find some cycles to submit some improvements.

**Comment by joestringer** [comment]
@aanm see #5162 for improved upgrade docs, I think that this should sufficiently document the above requirements?

**Comment by joestringer** [comment]
k8s builds timed out, saw also on another PR:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/5542/flowGraphTable/

Filed #5163 for the above.

Will rebase and re-run CI.

**Comment by joestringer** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Comment by joestringer** [comment]
Latest failure:

https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/5542/

We seem to lose chunks of time randomly:
```
...
13:59:30     k8s1-1.7: 701b4d847147: Download complete
14:06:10     k8s1-1.7: a221aee506c9: 
...
14:09:08     k8s1-1.7: 7f80f6cda1cc: 
14:09:08     k8s1-1.7: Pull complete
14:58:52     k8s1-1.7: 96649a2a4ebe: Verifying Checksum
...
15:03:56     k8s1-1.7: 9e65a62c4127: Download complete
15:19:41     k8s1-1.7: c64513b74145: 
...
15:19:54     k8s1-1.7: a98c60236860: Download complete
15:23:18     k8s1-1.7: 473d8689cf91: 
...
15:23:18     k8s1-1.7: Download complete
15:32:29     k8s1-1.7: 2e02b3453911: 
```

**Comment by joestringer** [comment]
test-me-please

**Review by borkmann — COMMENTED** [review]
Looks good. The RunGC() uses `bpf.OpenMap()` and then assumes that this has the same representation of key/value size as daemon internally uses in Go. Same `bpf.OpenMap()` would also be on `flushCtProto()` where we do a `ctmap.Flush()` and and walk it. `SyncMetricsMap()` is another candidate with potential upgrade/downgrade issues that could break given it's using `bpf.OpenMap()` as well and has expected Key/Value.

**Comment by jrfastab** [comment]
Removing backport-pending/1.0 and adding back needs-backport/1.0 I can't find a PR for this and its not in the 1.0 branch.

**Comment by jrfastab** [comment]
Link to discussion about documenting upgrade/downgrade fix with documentation instead of pulling multiple PRs into v1.0. #5118

**Comment by jrfastab** [comment]
Tracking fix in this issue,  https://github.com/cilium/cilium/issues/5454

**Pull request merged**